src/macros.rs

```diff
@@ -33,38 +33,209 @@ macro_rules! scopes {
     }};
 }
 
-/// Reduce boilerplate when inserting new elements in a JSON object.
+/// If there's an optional value, the macro will return its value. Otherwise, a
+/// default value will be returned. This is helpful to handle `$( expr )?`
+/// cases in macros.
 #[doc(hidden)]
 #[macro_export]
-macro_rules! json_insert {
-    ($json:expr, $p1:expr, $p2:expr) => {
-        $json
-            .as_object_mut()
-            .unwrap()
-            .insert($p1.to_string(), json!($p2))
+macro_rules! opt {
+    (, $default:expr) => {
+        $default
     };
+    ($optional:expr, $default:expr) => {
+        $optional
+    };
+}
+
+/// Count items in a list of items within a macro, taken from here:
+/// https://danielkeep.github.io/tlborm/book/blk-counting.html
+#[doc(hidden)]
+#[macro_export]
+macro_rules! replace_expr {
+    ($_t:tt $sub:expr) => {
+        $sub
+    };
+}
+#[doc(hidden)]
+#[macro_export]
+macro_rules! count_items {
+    ($($item:expr),*) => {<[()]>::len(&[$($crate::replace_expr!($item ())),*])};
+}
+
+/// Hack to convert an identifier to a string, including raw identifiers.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! ident_str {
+    ($name:ident) => {{
+        const NAME: &str = stringify!($name);
+        match &NAME.get(..2) {
+            Some("r#") => &NAME[2..],
+            _ => NAME,
+        }
+    }};
+}
+
+/// Private macro to insert either required or optional fields. Pattern matching
+/// will accordingly pick the branch, and then insert ($key, $val) into $map.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! params_internal {
+    ($map:ident, req, $name:ident, $key:expr, $val:expr) => {
```

Comment:
I don't really like that we have to write the entire word, as the macro is meant to have less code, but yes, I think it'll be easier to understand for a newcomer to the library, so it might be worth it. I've just updated it, let me know what you think.